File: sdk/streamlib-python-wheel/python/streamlib/audio_block.py
# Copyright (c) 2025 Jonathan Fontanez
# SPDX-License-Identifier: BUSL-1.1

"""The optional typed cast for an audio-block bag.

A link carries a self-describing bag (a dict); its keys are the contract and
reading them directly is always enough. Casting is the opt-in dial for
consumers that want construction-time validation and attribute access instead
of key lookups.

Unlike a video frame, an audio block carries its payload inline: the samples
ride the bag as msgpack ``bin``, which reaches Python as ``bytes``, and
``dtype`` says how to read those bytes. So this cast composes nothing
surface-shaped — there is no surface id, no claim and no lifetime to hold, and
``ClaimedSurfacePixelAccess`` demands all three of a type that composes it.

What ``samples`` guarantees is that reading the block adds no copy of its
payload: the numpy array is a view over the bag's own ``bytes``. The copies
between shared memory and ``process()`` are the helper hop every bag pays and
this cast removes none of them.
"""

from __future__ import annotations

import typing
from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Any, Literal

if typing.TYPE_CHECKING:
    from numpy.typing import NDArray

__all__ = ["AudioBlock"]

AudioSampleDtype = Literal["f32", "i16"]

# Keyed by the same wire vocabulary: how to read one scalar, and how wide it
# is. The numpy spelling is explicitly little-endian — never the platform's
# native one — because the payload is little-endian by contract and not by
# luck, and a big-endian reader taking the native spelling would decode
# every sample wrong rather than fail.
_NUMPY_TYPE_FOR_DTYPE: "dict[str, str]" = {"f32": "<f4", "i16": "<i2"}
_BYTES_PER_SAMPLE_FOR_DTYPE: "dict[str, int]" = {"f32": 4, "i16": 2}

# The keys without which there is no block to speak of. Everything else the
# bag carries is optional, this cast's or somebody else's.
_REQUIRED_BAG_KEYS = (
    "samples",
    "sample_rate",
    "channels",
    "sample_count",
    "first_sample_timestamp_ns",
)


def _is_plain_int(value: Any) -> bool:
    """`bool` is an `int` subclass; a channel count of `True` is a bug."""
    return isinstance(value, int) and not isinstance(value, bool)


@dataclass(frozen=True, init=False)
class AudioBlock:
    """An audio-block bag, cast: interleaved CPU samples plus how to read them.

    ``first_sample_timestamp_ns`` (the machine's monotonic clock in
    nanoseconds, comparable across every process on the host) is the ordering
    primitive and the whole of A/V sync: any sample's instant derives from it,
    ``sample_count`` and ``sample_rate``, so joining a block to a camera frame
    is subtracting two timestamps.

    Read through ``ctx.inputs.read(port, into=AudioBlock)``, whose bag keys
    this validates on construction; ``samples`` is then the numpy view of the
    payload.
    """

    # Off the repr: a block is thousands of scalars, and a failed assertion
    # that prints all of them buries what it was actually about.
    interleaved_sample_bytes: bytes = field(repr=False)
    sample_rate: int
    channels: int
    sample_count: int
    dtype: AudioSampleDtype
    first_sample_timestamp_ns: int

    def __init__(
        self,
        samples: bytes,
        sample_rate: int,
        channels: int,
        sample_count: int,
        first_sample_timestamp_ns: int,
        dtype: AudioSampleDtype = "f32",
        **keys_this_cast_does_not_read: Any,
    ) -> None:
        """Validate the bag's entries and assign them.

        Written out rather than generated because the bag is an open map: a
        producer may carry keys this cast does not read, and the day one adds
        a key must not be the day every `read(port, into=AudioBlock)` starts
        raising. Construction is the validation, so that spelling and
        `from_bag` are one path rather than two that can drift.
        """
        del keys_this_cast_does_not_read
        if not isinstance(samples, bytes):
            raise ValueError(
                "bag is not an audio block: 'samples' must be bytes — the payload "
                "rides the bag as msgpack bin"
            )
        if (
            not _is_plain_int(sample_rate)
            or not _is_plain_int(channels)
            or not _is_plain_int(sample_count)
            or not _is_plain_int(first_sample_timestamp_ns)
        ):
            raise ValueError(
                "bag is not an audio block: sample_rate/channels/sample_count/"
                "first_sample_timestamp_ns must be int"
            )
        # Before the length check, which two negatives would slip through by
        # cancelling: a payload of four bytes satisfies
        # `sample_count=-1 × channels=-1 × 4`, and the block would then fail at
        # `reshape` rather than here. The Rust cast spells these `u32`, where
        # the case cannot arise at all.
        if sample_count < 0 or channels < 0:
            raise ValueError(
                f"bag is not an audio block: sample_count={sample_count} and "
                f"channels={channels} must both be non-negative"
            )
        if dtype not in _NUMPY_TYPE_FOR_DTYPE:
            raise ValueError(
                f"bag is not an audio block: dtype {dtype!r} is not one this cast "
                f"reads ({', '.join(sorted(_NUMPY_TYPE_FOR_DTYPE))})"
            )
        # Checked here rather than left to `reshape`, which would take a
        # payload of the wrong length and either raise about shapes or — when
        # the length happens to divide — hand back a plausible-looking wrong
        # answer.
        bytes_per_sample = _BYTES_PER_SAMPLE_FOR_DTYPE[dtype]
        expected_byte_count = sample_count * channels * bytes_per_sample
        if len(samples) != expected_byte_count:
            raise ValueError(
                f"bag is not an audio block: 'samples' carries {len(samples)} bytes, "
                f"but sample_count={sample_count} × channels={channels} × "
                f"{bytes_per_sample} bytes per {dtype} sample is {expected_byte_count}"
            )
        object.__setattr__(self, "interleaved_sample_bytes", samples)
        object.__setattr__(self, "sample_rate", sample_rate)
        object.__setattr__(self, "channels", channels)
        object.__setattr__(self, "sample_count", sample_count)
        object.__setattr__(self, "dtype", dtype)
        object.__setattr__(
            self, "first_sample_timestamp_ns", first_sample_timestamp_ns
        )

    @property
    def samples(self) -> "NDArray[Any]":
        """The payload as a ``(sample_count, channels)`` numpy array.

        A view over ``interleaved_sample_bytes``, so it costs no copy of the
        payload and is read-only — the bag's bytes are immutable. Take a copy
        (``numpy.array(block.samples)``) to write into it.
        """
        # Imported here so the wheel never takes a numpy dependency: a user
        # reaching for the array already has numpy.
        import numpy

        return numpy.frombuffer(
            self.interleaved_sample_bytes, dtype=_NUMPY_TYPE_FOR_DTYPE[self.dtype]
        ).reshape(self.sample_count, self.channels)

    @classmethod
    def from_bag(cls, bag: Mapping[str, Any]) -> "AudioBlock":
        """Construct from a bag dict, raising ValueError on a missing or
        mistyped key.

        The same *validation* `read(port, into=AudioBlock)` performs, so the
        two spellings cannot disagree about what a valid block is; this one
        only names the missing key, which keyword construction reports in
        Python's words.
        """
        missing = [key for key in _REQUIRED_BAG_KEYS if key not in bag]
        if missing:
            raise ValueError(f"bag is not an audio block: missing key {missing[0]!r}")
        return cls(**bag)

File: runtime/streamlib-media-builtins/src/audio_block.rs
// Copyright (c) 2025 Jonathan Fontanez
// SPDX-License-Identifier: BUSL-1.1

//! The audio-block bag convention the built-ins produce and consume.
//!
//! A link carries a self-describing msgpack named map; these types are the
//! optional Rust cast for it — never declared on a port, never registered
//! anywhere. The field names ARE the wire contract: a consumer in any
//! language reads the same keys from the bag dict. The map is open — a
//! producer may carry extra keys and this cast ignores them, matching the
//! Python cast's behavior.
//!
//! Unlike a video frame, an audio block carries its payload inline: the
//! samples ride the bag as msgpack `bin`, and `dtype` says how to read those
//! bytes.

use serde::{Deserialize, Serialize};

/// Audio block bag: interleaved CPU samples ride the link inline, described
/// by the rate, channel count, and dtype beside them.
///
/// `first_sample_timestamp_ns` is the ordering primitive and the whole of
/// A/V sync: any sample's instant derives from it, `sample_count` and
/// `sample_rate`, so joining a block to a camera frame is subtracting two
/// timestamps in the same monotonic epoch.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct AudioBlock {
    /// The block's scalars, interleaved by channel, little-endian, read
    /// according to `dtype`.
    ///
    /// The wire key is `samples` and the payload is msgpack `bin`: one field
    /// spelling serves every dtype, and little-endian is a wire statement
    /// rather than an assumption — it is what a tap, a CLI, or a consumer in
    /// another language depends on.
    #[serde(rename = "samples", with = "serde_bytes")]
    pub interleaved_sample_bytes: Vec<u8>,
    /// Sample rate in hertz.
    pub sample_rate: u32,
    /// Channel count the payload is interleaved by.
    pub channels: u32,
    /// Per-channel sample count: the payload carries
    /// `sample_count × channels` scalars, and the block's duration is
    /// `sample_count / sample_rate`.
    pub sample_count: u32,
    /// How to read `samples`. Absent on the wire means `f32`.
    #[serde(default)]
    pub dtype: AudioSampleDtype,
    /// Monotonic timestamp in nanoseconds of the block's first sample,
    /// stamped by the capturing device — the machine's monotonic epoch, the
    /// one a `VideoFrame.timestamp_ns` is stamped in.
    pub first_sample_timestamp_ns: i64,
}

/// How the scalars in an [`AudioBlock`]'s payload are encoded.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum AudioSampleDtype {
    #[default]
    #[serde(rename = "f32")]
    F32,
    #[serde(rename = "i16")]
    I16,
}

impl AudioSampleDtype {
    /// Bytes one scalar occupies in the payload.
    pub fn bytes_per_sample(self) -> usize {
        match self {
            AudioSampleDtype::F32 => 4,
            AudioSampleDtype::I16 => 2,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn interleaved_f32_bytes(scalars: &[f32]) -> Vec<u8> {
        scalars.iter().flat_map(|scalar| scalar.to_le_bytes()).collect()
    }

    fn interleaved_i16_bytes(scalars: &[i16]) -> Vec<u8> {
        scalars.iter().flat_map(|scalar| scalar.to_le_bytes()).collect()
    }

    fn wire_map_entry(wire_bytes: &[u8], key_name: &str) -> rmpv::Value {
        let value: rmpv::Value =
            rmpv::decode::read_value(&mut &wire_bytes[..]).expect("msgpack decode");
        let rmpv::Value::Map(entries) = value else {
            panic!("wire value must be a named map, got {value:?}");
        };
        entries
            .iter()
            .find(|(key, _)| key.as_str() == Some(key_name))
            .unwrap_or_else(|| panic!("wire map missing key {key_name:?}"))
            .1
            .clone()
    }

    /// The actual wire is msgpack via `rmp_serde::to_vec_named` (what
    /// `OutputWriter::write` does), and `samples` has to land on it as `bin`.
    ///
    /// Asserted as the msgpack type rather than through a serde round trip,
    /// which passes either way: a `Vec<f32>` field would encode as a msgpack
    /// *array* — five bytes per sample, and a shape another language reads as
    /// a list of numbers rather than a buffer.
    #[test]
    fn audio_block_msgpack_wire_carries_the_samples_as_a_binary_payload() {
        let block = AudioBlock {
            interleaved_sample_bytes: interleaved_f32_bytes(&[
                -1.0, -0.5, 0.0, 0.25, 0.5, 1.0,
            ]),
            sample_rate: 48_000,
            channels: 2,
            sample_count: 3,
            dtype: AudioSampleDtype::F32,
            first_sample_timestamp_ns: 123_456_789,
        };

        let wire_bytes = rmp_serde::to_vec_named(&block).expect("msgpack serialize");
        assert_eq!(
            wire_map_entry(&wire_bytes, "samples"),
            rmpv::Value::Binary(block.interleaved_sample_bytes.clone()),
            "the samples field is a byte buffer, not a sequence of scalars"
        );
        assert_eq!(
            wire_map_entry(&wire_bytes, "sample_rate").as_u64(),
            Some(48_000)
        );
        assert_eq!(wire_map_entry(&wire_bytes, "channels").as_u64(), Some(2));
        assert_eq!(wire_map_entry(&wire_bytes, "sample_count").as_u64(), Some(3));
        assert_eq!(wire_map_entry(&wire_bytes, "dtype").as_str(), Some("f32"));
        assert_eq!(
            wire_map_entry(&wire_bytes, "first_sample_timestamp_ns").as_i64(),
            Some(123_456_789)
        );

        let round_tripped: AudioBlock =
            rmp_serde::from_slice(&wire_bytes).expect("msgpack deserialize");
        assert_eq!(round_tripped, block);
    }

    /// One field spelling serves every dtype: an `i16` block's payload is the
    /// same `bin`, half the width per scalar.
    #[test]
    fn an_i16_block_carries_its_samples_as_a_binary_payload_too() {
        let block = AudioBlock {
            interleaved_sample_bytes: interleaved_i16_bytes(&[i16::MIN, -1, 0, i16::MAX]),
            sample_rate: 16_000,
            channels: 1,
            sample_count: 4,
            dtype: AudioSampleDtype::I16,
            first_sample_timestamp_ns: 7,
        };

        let wire_bytes = rmp_serde::to_vec_named(&block).expect("msgpack serialize");
        assert_eq!(
            wire_map_entry(&wire_bytes, "samples"),
            rmpv::Value::Binary(block.interleaved_sample_bytes.clone())
        );
        assert_eq!(wire_map_entry(&wire_bytes, "dtype").as_str(), Some("i16"));
        assert_eq!(
            block.interleaved_sample_bytes.len(),
            block.sample_count as usize
                * block.channels as usize
                * block.dtype.bytes_per_sample(),
            "an interleaved block carries sample_count × channels scalars"
        );

        let round_tripped: AudioBlock =
            rmp_serde::from_slice(&wire_bytes).expect("msgpack deserialize");
        assert_eq!(round_tripped, block);
    }

    /// The bag map is open: a producer carrying extra keys must not break
    /// this cast (mirrors the Python cast's behavior).
    #[test]
    fn audio_block_cast_ignores_unknown_keys() {
        let wire_bytes = wire_bytes_for(vec![
            ("samples", rmpv::Value::Binary(interleaved_f32_bytes(&[0.5]))),
            ("sample_rate", rmpv::Value::from(44_100)),
            ("channels", rmpv::Value::from(1)),
            ("sample_count", rmpv::Value::from(1)),
            ("dtype", rmpv::Value::from("f32")),
            ("first_sample_timestamp_ns", rmpv::Value::from(11)),
            ("a_future_key", rmpv::Value::from("ignored")),
        ]);

        let block: AudioBlock = rmp_serde::from_slice(&wire_bytes).expect("open map");
        assert_eq!(block.sample_rate, 44_100);
        assert_eq!(block.interleaved_sample_bytes, interleaved_f32_bytes(&[0.5]));
    }

    /// `dtype` is metadata with a default, so a producer that omits it is
    /// describing an `f32` block rather than an undecodable one.
    #[test]
    fn a_block_with_no_dtype_on_the_wire_reads_as_f32() {
        let wire_bytes = wire_bytes_for(vec![
            ("samples", rmpv::Value::Binary(interleaved_f32_bytes(&[1.0]))),
            ("sample_rate", rmpv::Value::from(8_000)),
            ("channels", rmpv::Value::from(1)),
            ("sample_count", rmpv::Value::from(1)),
            ("first_sample_timestamp_ns", rmpv::Value::from(0)),
        ]);

        let block: AudioBlock = rmp_serde::from_slice(&wire_bytes).expect("dtype defaults");
        assert_eq!(block.dtype, AudioSampleDtype::F32);
    }

    fn wire_bytes_for(entries: Vec<(&str, rmpv::Value)>) -> Vec<u8> {
        let map = rmpv::Value::Map(
            entries
                .into_iter()
                .map(|(key, value)| (rmpv::Value::from(key), value))
                .collect(),
        );
        let mut wire_bytes = Vec::new();
        rmpv::encode::write_value(&mut wire_bytes, &map).expect("msgpack encode");
        wire_bytes
    }
}
